src/llmq/quorums_blockprocessor.h

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2018 The Dash Core developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef DASH_QUORUMS_BLOCKPROCESSOR_H
+#define DASH_QUORUMS_BLOCKPROCESSOR_H
+
+#include "llmq/quorums_commitment.h"
+
+#include "consensus/params.h"
+#include "primitives/transaction.h"
+#include "sync.h"
+
+#include <map>
+
+class CNode;
+class CConnman;
+
+namespace llmq
+{
+
+class CQuorumBlockProcessor
+{
+private:
+    CEvoDB& evoDb;
+
+    // TODO cleanup
+    CCriticalSection minableCommitmentsCs;
+    std::map<std::pair<Consensus::LLMQType, uint256>, uint256> minableCommitmentsByQuorum;
+    std::map<uint256, CFinalCommitment> minableCommitments;
+
+public:
+    CQuorumBlockProcessor(CEvoDB& _evoDb) : evoDb(_evoDb) {}
+
+    void ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStream& vRecv, CConnman& connman);
+
+    bool ProcessBlock(const CBlock& block, const CBlockIndex* pindexPrev, CValidationState& state);
+    bool UndoBlock(const CBlock& block, const CBlockIndex* pindex);
+
+    void AddMinableCommitment(const CFinalCommitment& fqc);
+    bool HasMinableCommitment(const uint256& hash);
+    bool GetMinableCommitmentByHash(const uint256& commitmentHash, CFinalCommitment& ret);
+    bool GetMinableCommitment(Consensus::LLMQType llmqType, const CBlockIndex* pindexPrev, CFinalCommitment& ret);
+    bool GetMinableCommitmentTx(Consensus::LLMQType llmqType, const CBlockIndex* pindexPrev, CTransactionRef& ret);
+
+    bool HasMinedCommitment(Consensus::LLMQType llmqType, const uint256& quorumHash);
+    bool GetMinedCommitment(Consensus::LLMQType llmqType, const uint256& quorumHash, CFinalCommitment& ret);
+
+private:
+    bool GetCommitmentsFromBlock(const CBlock& block, const CBlockIndex* pindexPrev, std::map<Consensus::LLMQType, CFinalCommitment>& ret, CValidationState& state);
+    bool ProcessCommitment(const CBlockIndex* pindexPrev, const CFinalCommitment& qc, CValidationState& state);
+    bool IsMiningPhase(Consensus::LLMQType llmqType, int nHeight);
+    uint256 GetQuorumBlockHash(Consensus::LLMQType llmqType, const CBlockIndex* pindexPrev);
+};
+
+extern CQuorumBlockProcessor* quorumBlockProcessor;
```

Comment:
I know we have other places w/globals like this (e.g. sporkManager), but is there any way to avoid them in the new code?

Reply:
How? We need these object everywhere, so we need some type of global/singleton instance. The alternative is to put these into static class members, but IMHO this is in no way better then global variables. One big downsides of global variables is the cluttering of the global namespace, but this is only partially the case here as these live in their own namespace (`llmq`)